### ISSUE TITLE
Why MultiImageInput (from master) forces you to wrap API return value with a list?

### ISSUE THREAD

**Issue by guy4261** [bentoml/BentoML#883]
Hello,

I `pip install -e .` the latest `master` branch, in order to use the `MultiImageInput` adapter.

[Looking at the code](https://github.com/bentoml/BentoML/blob/master/bentoml/adapters/multi_image_input.py#L101-L107), I see:

```python
    def handle_request(self, request: Request, func):
        ...
        result = func((files,))[0]
        return self.output_adapter.to_response(result, request)
```

Why is `[0]` applied to `func((files,))`? It forces me to also return an object wrapped in a list:

```python
    @bentoml.api(input=MultiImageInput())
    def predict(self, imgs):
        ...
        return [ret]
```

Which is something I can do - but I don't understand the logic behind it, and not sure it was intentional (as this feature was not yet officially released). So I thought I'd ask.

Thanks in advance!

**Comment by parano** [comment]
Hey @guy4261  -  related documents for this are coming out soon. basically the user-defined API functions are expected to take a list of inference inputs and return a list of outputs(see 
https://docs.bentoml.org/en/latest/concepts.html#api-function-and-adapters). This makes it possible to do adaptive micro-batching, which brings a huge performance advantage to your online serving workloads.

**Comment by guy4261** [comment]
Thanks @parano !

I understand that to support microbatching, a `@bentoml.api` accepts and returns a list of predictions.

But if `ret` in my example is a list, what I would get back is just the first item (because of the `[0]` [here](https://github.com/bentoml/BentoML/blob/master/bentoml/adapters/multi_image_input.py#L101-L107)). I tried that and that's the way this works right now.

So why
`result = func((files,))[0]`   
instead of
`result = func((files,))`  
?

**Comment by guy4261** [comment]
I get it now - `files` are wrapped in a tuple (hence `(files,)`), and likewise the result should be wrapped in a list/tuple.
But if you wrap `files` in a tuple, then take the first results `[0]` of the result,
wouldn't it be simpler to do neither of these?

How does wrapping exactly one item / unwrapping the results (taking the first or `[0]` result) supports microbatching?

**Comment by parano** [comment]
From the user's perspective, it's just implementing the API function and make sure it takes a list of inputs, and return a list of outputs, and make sure they have the same length and order.

For example, if your model takes a JSON list as input:

```
def
```

**Comment by guy4261** [comment]
OK - got it! I got back to another api I had on this bento which has `JsonInput`. What worked when I `pip install bentoml` does not work from the editable install of `master`. So now I realize the change.

So now the API is that of microbatching (even if it's not enabled btw?) and the API endpoints receive batches (=lists) of inputs:
* `JsonInput` receives a list of JSONs,
* `MultiImageInput` receives a list of multi-images (which are themselves lists of images).

and the framework is responsible for merging (albeit order+length might change so it's the user's job to verify everything got back properly).

Thanks! If I got it correctly, feel free to say so and close this issue 🙏 .

In that sense btw - if I have 10 images I want predictions for, what's better? To send 1 request and use the `MultiImageHandler`? Or to send 10 requests, each one for a single image?
